File: antia_numerical_methods@Beliavsky/f77/xmp8/pcor.f
!       Calculate the correlation coefficient and the probability of
!       finding a value as high as that for uncorrelated dats sets

      PROGRAM XYCOR
      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION X(1000),Y(1000)

51    FORMAT(/,'  N =',I5,'  Correlation Coefficient =',F8.4,/,
     1  'Probability of finding this value or higher for uncorrelated',
     2  ' sequence =',1PD12.3)

      SEED=9
100   PRINT *,'TYPE N = NO. OF DATA POINTS (QUITS WHEN N.LE.0)'
      READ *,N
      IF(N.LE.0.OR.N.GT.1000) STOP
!       Generate the data
      DO I=1,N
        R=RAN1(SEED)
        X(I)=I*1D-2+1.D-3*(R-0.5)
        R=RAN1(SEED)
        Y(I)=I*0.001D-2+1.D-2*(R-0.5)
      ENDDO

!       Calculate the correlation coefficient between X and Y
      XMEAN=0.0
      YMEAN=0.0
      DO I=1,N
        XMEAN=XMEAN+X(I)
        YMEAN=YMEAN+Y(I)
      ENDDO
      XMEAN=XMEAN/N
      YMEAN=YMEAN/N
      SXX=0.0
      SYY=0.0
      SXY=0.0
      DO I=1,N
        SXX=SXX+(X(I)-XMEAN)**2
        SYY=SYY+(Y(I)-YMEAN)**2
        SXY=SXY+(X(I)-XMEAN)*(Y(I)-YMEAN)
      ENDDO
      RCOR=SXY/SQRT(SXX*SYY)
      PROB=PCOR(N-2,ABS(RCOR))
      WRITE(6,51) N,RCOR,PROB
      GO TO 100
      END

!       -----------------------------------------------------

!	To calculate the Error function for a real argument
!
!	Required routines : None
 
      FUNCTION ERF(X)
      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION A(7),B(8),A1(7),B1(8)

!	The coefficients of rational function approximations
      DATA A/ 4.891837297874833514D-01,  1.110175596090841322D-01,
     1        1.526977188817169289D-02,  1.388143322498740953D-03,
     1        8.446154421878829637D-05,  3.239915935631695227D-06,
     1        6.200069065781009292D-08/
      DATA B/ 1.128379167095512575D+00,  1.758583405424390318D-01,
     1        5.411290829613803886D-02,  3.805760982281134630D-03,
     1        4.314620532020313106D-04,  1.423201530735308681D-05,
     1        6.188698151904667110D-07,  2.293915472550064153D-09/
      DATA A1/ 3.040844316651787557D+01,  3.358927547920980388D+02,
     1         1.703170048554673596D+03,  4.133195606956137156D+03,
     1         4.555611776312694034D+03,  1.935778559646575488D+03,
     1         2.051089518116697193D+02/
      DATA B1/ 5.641895835477562749D-01,  1.687403209467950089D+01,
     1         1.813522721872712655D+02,  8.779664433851135986D+02,
     1         1.965115658619443782D+03,  1.865558781108286245D+03,
     1         5.828865035607128761D+02,  2.558559157228883880D+01/
 
      IF(ABS(X).LT.2.0) THEN
!	Use approximation for Erf(x)/x
        Y=X*X
        FN=((((((B(8)*Y+B(7))*Y+B(6))*Y+B(5))*Y+B(4))*Y+B(3))*Y+B(2))*Y
     1      +B(1)
        FD=((((((A(7)*Y+A(6))*Y+A(5))*Y+A(4))*Y+A(3))*Y+A(2))*Y
     1      +A(1))*Y+1
        ERF=X*FN/FD
      ELSE
 
!	Use approximation for x*Exp(x^2)*Erfc(x)
        X1=1./X**2
        FN=((((((B1(8)*X1+B1(7))*X1+B1(6))*X1+B1(5))*X1+B1(4))*X1+
     1      B1(3))*X1+B1(2))*X1+B1(1)
        FD=((((((A1(7)*X1+A1(6))*X1+A1(5))*X1+A1(4))*X1+A1(3))*X1+
     1      A1(2))*X1+A1(1))*X1+1
        ERF=1.-EXP(-X*X)*FN/(FD*ABS(X))
        IF(X.LT.0.0) ERF=-ERF
      ENDIF
 
      END

!       -----------------------------------------------------

!	To calculate the Logarithm of Gamma function for a real argument
!	For negative values it give ln(abs(Gamma(x)))
!
!	Required routines : None
 
      FUNCTION GAMMAL(X)
      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION A(4),B(4),A1(6),B1(6)
!	PI2L=LOG(2*PI)/2
      PARAMETER (PI2L=0.918938533204672741D0,PI=3.141592653589793238D0)

!	The coefficients of rational function approximations
      DATA A/ 1.013142782275024216D-2,  7.645657825398191944D-1,
     1        3.381172379819461227D-4,  1.595363637547538209D-2/
      DATA B/0.0, 8.333333333338911768D-2, 8.442856404442060242D-4,
     1            6.093603832366013704D-2/

      DATA A1/ 4.681163846241230144D0,  3.208225429683256526D0,
     1         5.145525793448859216D-1, 1.581117883959157936D-2,
     2        -6.398416804905407512D-5, 5.264566254181773919D-7/
      DATA B1/ 2.938038561191284576D0,  1.489364948862436743D0,
     1        -5.466291543917642961D0,  1.972497734170110410D-1,
     2         7.830146473241555157D-1, 5.756753067834747499D-2/
 
      T=ABS(X)
      IT=T
      IF(T.GE.10.0) THEN
!	Use asymptotic approximation for T>10
        Y=1/T
        FN=((B(4)*Y+B(3))*Y+B(2))*Y+B(1)
        FD=(((A(4)*Y+A(3))*Y+A(2))*Y+A(1))*Y+1
        GAMMAL=FN/FD-T+PI2L+(T-0.5)*LOG(T)
      ELSE
 
!	Use approximation for [4,5]
        FAC=0.0
        T1=T
        IF(T.LT.4) THEN
          DO I=IT,3
            FAC=FAC-LOG(T1)
            T1=T1+1
          ENDDO
        ELSE IF(T.GT.5) THEN
          DO I=IT,5,-1
            T1=T1-1
            FAC=FAC+LOG(T1)
          ENDDO
        ENDIF
        FN=((((B1(6)*T1+B1(5))*T1+B1(4))*T1+B1(3))*T1+B1(2))*T1+B1(1)
        FD=(((((A1(6)*T1+A1(5))*T1+A1(4))*T1+A1(3))*T1+A1(2))*T1+
     1          A1(1))*T1+1
        GAMMAL=FN/FD+FAC
      ENDIF

      IF(X.LE.0) THEN
        IF(T.GT.IT) THEN
          GAMMAL=LOG(PI)-LOG(ABS(SIN(PI*T)*T))-GAMMAL
        ELSE
          GAMMAL=(-1)**IT/0.0
        ENDIF
      ENDIF
 
      END

!       -----------------------------------------------------

!	To calculate the probability that two uncorrelated sequences
!	of length (n+2) will give a correlation coefficient exceeding |x|
!	For large even values of n the series can give large roundoff
!	in which case the distribution is approximated by Normal distribution
!
!       N  : (input) Length of data sets should be N+2
!       XX : (input) The function will calculate the probability for
!               correlation exceeding |x| for uncorrelated sequences
!
!	Required routines : GAMMAL, ERF

      FUNCTION PCOR(N,XX)
      IMPLICIT REAL*8(A-H,O-Z)
      PARAMETER(PI=3.14159265358979323846D0,PIS=1.7724538509055144D0)

      X=ABS(XX)
      PCOR=0.0
      IF(X.GT.1.0) RETURN
      IF(2*(N/2).EQ.N) THEN
!	if n is even
        L=(N-2)/2
        SS=X
        P=X
        PMAX=P
        DO K=1,l
          P=-(L-K+1)*P*X*X/K
          SS=SS+P/(2*K+1.D0)
          IF(ABS(P).GT.PMAX) PMAX=ABS(P)
        ENDDO
        PL=GAMMAL((N+1.D0)/2)-GAMMAL(N/2.D0)
        PCOR=2.*EXP(PL)*SS/PIS
        IF(PMAX.GT.1.D5.OR.PCOR.GT.1.0) PCOR=ERF(X*SQRT(N/2.D0))
      ELSE
!	if n is odd
        L=(N-3)/2
        SS=SQRT(1.-X*X)
        P=SS
        IF(N.EQ.1) SS=0.0
        DO K=1,L
          P=P*(2*K)*(1-X*X)/(2*K+1.D0)
          SS=SS+P
        ENDDO
        PCOR=(ASIN(X)+X*SS)*2/PI
      ENDIF
      PCOR=1-PCOR
      END

!       -----------------------------------------------------

!	To generate uniformly distributed random numbers in interval (0,1)
!
!	SEED : (input/output) is a real value used as the seed
!		It should be positive during initial call and
!		should not be modified between different calls.
!
!	Required routines : None

      FUNCTION RAN1(SEED)
      IMPLICIT REAL*8(A-H,O-Z)
!	Retain the following declaration even for REAL*4 version
!	otherwise AM and AC will be rounded
      REAL*8 AM,A,AC,AN
      PARAMETER(AM=2147483648D0,A=45875,AC=453816693D0,AN=2147483647D0)

      SEED=MOD(SEED*A+AC,AM)
      RAN1=SEED/AN
      END

File: projects/antia_numerical_methods@Beliavsky/c/xmp/pcor.c
/*	Calculate the correlation coefficient and the probability of
        finding a value as high as that for uncorrelated dats sets */


#include <stdio.h>
#include <math.h>

double gammaln(double xg);
double erf(double x);
double pcor(int n, double xx);
double ran1(double *seed);


main()
{
	int i,n;
	double seed,r,xmean,ymean,sxx,syy,sxy,rcor,prob,rx,ry;
	double x[100],y[100];

	seed=9; n=100;
/*	Generate the data */
	for(i=0; i<n; ++i) {
		r=ran1(&seed); x[i]=i*0.01+0.001*(r-0.5);
		r=ran1(&seed); y[i]=i*1e-5+0.01*(r-0.5);
	}

/*	Calculate the correlation coefficient between X and Y */
	xmean=0.0; ymean=0.0;
	for(i=0; i<n; ++i){
		xmean=xmean+x[i]; ymean=ymean+y[i];}
	xmean=xmean/n; ymean=ymean/n;
	sxx=0.0; syy=0.0; sxy=0.0;
	for(i=0; i<n; ++i){
		rx=x[i]-xmean; ry=y[i]-ymean;
		sxx=sxx+rx*rx; syy=syy+ry*ry; sxy=sxy+rx*ry;
	}
	rcor=sxy/sqrt(sxx*syy);
	prob=pcor(n-2,fabs(rcor));
	printf("Correlation Coefficient = %f\n",rcor);
	printf("Probability of finding this value or higher for uncorrelated sequence =,%le\n",prob);
}

/*	To calculate the Error function for a real argument 

	Required functions : None
*/

#include <math.h>

double erf(double x)

{
	double fn,fd,f,y;

/*	The coefficients of rational function approximations */
	double a[7] = {4.891837297874833514e-01,  1.110175596090841322e-01,
                       1.526977188817169289e-02,  1.388143322498740953e-03,
                       8.446154421878829637e-05,  3.239915935631695227e-06,
                       6.200069065781009292e-08};
	double b[8] = {1.128379167095512575e+00,  1.758583405424390318e-01,
                       5.411290829613803886e-02,  3.805760982281134630e-03,
                       4.314620532020313106e-04,  1.423201530735308681e-05,
                       6.188698151904667110e-07,  2.293915472550064153e-09};
	double a1[7] = {3.040844316651787557e+01,  3.358927547920980388e+02,
                        1.703170048554673596e+03,  4.133195606956137156e+03,
                        4.555611776312694034e+03,  1.935778559646575488e+03,
                        2.051089518116697193e+02};
	double b1[8] = {5.641895835477562749e-01,  1.687403209467950089e+01,
                        1.813522721872712655e+02,  8.779664433851135986e+02,
                        1.965115658619443782e+03,  1.865558781108286245e+03,
                        5.828865035607128761e+02,  2.558559157228883880e+01};
 
	if(fabs(x)<2.0) {
/*	Use approximation for Erf(x)/x */
		y=x*x;
		fn=((((((b[7]*y+b[6])*y+b[5])*y+b[4])*y+b[3])*y+b[2])*y+b[1])*y+b[0];
		fd=((((((a[6]*y+a[5])*y+a[4])*y+a[3])*y+a[2])*y+a[1])*y+a[0])*y+1;
		f=x*fn/fd;
	}

	else {
/*	Use approximation for x*Exp(x^2)*Erfc(x) */
		y=1.0/(x*x);
		fn=((((((b1[7]*y+b1[6])*y+b1[5])*y+b1[4])*y+b1[3])*y+b1[2])*y+b1[1])*y+b1[0];
		fd=((((((a1[6]*y+a1[5])*y+a1[4])*y+a1[3])*y+a1[2])*y+a1[1])*y+a1[0])*y+1;
		f=1.-exp(-x*x)*fn/(fd*fabs(x));
		if(x<0.0) f=-f;
	}
	return f;
}


/*	To calculate the Logarithm of Gamma function for a real argument
 	For negative values it give ln(abs(Gamma(x)))
 
 	Required routines : None
*/

#include <math.h>

#define PI 3.14159265358979323846

double gammaln(double xg)

{
	double x,gx,f,y,rmk,f1,x1,fn,fd;
/*	PIL is LOG(SQRT(2*PI)) */
	double PI2L=0.918938533204672741;
	
/*	The coefficients for rational function approximations */
	double a[4] = {1.013142782275024216e-2,  7.645657825398191944e-1,
             3.381172379819461227e-4,  1.595363637547538209e-2};
	double b[4] ={0.0, 8.333333333338911768e-2, 8.442856404442060242e-4,
                 6.093603832366013704e-2};
	double a1[6] ={4.681163846241230144,  3.208225429683256526,
              5.145525793448859216e-1, 1.581117883959157936e-2,
             -6.398416804905407512e-5, 5.264566254181773919e-7};
	double b1[6] ={2.938038561191284576,  1.489364948862436743,
             -5.466291543917642961,  1.972497734170110410e-1,
              7.830146473241555157e-1, 5.756753067834747499e-2};
 
	x=fabs(xg);
	if(x>10.0) {
/*	Use asymptotic approximation (Stirling formula) */
		y=1.0/x;
		fn=((b[3]*y+b[2])*y+b[1])*y+b[0];
		fd=(((a[3]*y+a[2])*y+a[1])*y+a[0])*y+1;
		f=(x-0.5)*log(x)-x+PI2L+fn/fd;
	}

	else {
/*	Use rational function approximation for (Gamma) over [4,5]
	after translating the range if necessary */
		f1=1.0; x1=x;
		while(x1>5.0) {
			f1=f1*(x1-1.0);
			x1=x1-1.0;
		}
		while(x1<4.0) {
			f1=f1/(x1);
			x1=x1+1.0;
		}


		fn=((((b1[5]*x1+b1[4])*x1+b1[3])*x1+b1[2])*x1+b1[1])*x1+b1[0];
		fd=(((((a1[5]*x1+a1[4])*x1+a1[3])*x1+a1[2])*x1+a1[1])*x1+a1[0])*x1+1;
		f=log(f1)+fn/fd;
	}

	if(xg>0.0) return f;
	f=log(PI/(x*fabs(sin(PI*x))))-f;
	return f;
}
 
/*	To calculate the probability that two uncorrelated sequences
	of length (n+2) will give a correlation coefficient exceeding |x|
	For large even values of n the series can give large roundoff
	in which case the distribution is approximated by Normal distribution

	N  : (input) Length of data sets should be N+2
        XX : (input) The function will calculate the probability for
                correlation exceeding |x| for uncorrelated sequences

	Required routines : GAMMALN, ERF
*/

#include <math.h>

double gammaln(double xg);
double erf(double x);

#	define PI 3.14159265358979323846

double pcor(int n, double xx)

{
	double f,p,ss,pl,pmax,x;
/*	pis is sqrt(pi) */
	double pis=1.7724538509055144;
	int l,k;

	x=fabs(xx);
	if(x>1.0) return 0.0;
	if(2*(n/2)==n)
	{
/*	if n is even */
		l=(n-2)/2; ss=x; p=x; 
		pmax=p;
		for(k=1; k<=l; k++)
		{
			p=-(l-k+1)*p*x*x/k; ss=ss+p/(2*k+1.0);
			if(fabs(p)>pmax) pmax=fabs(p);
		}
		pl=gammaln((n+1.0)/2)-gammaln(n/2.0);
		f= 2.*exp(pl)*ss/pis;
		if(pmax>1.e5||f>1.0) f=erf(x*sqrt(n/2.0));
	}
	else
	{
/*	if n is odd */
		l=(n-3)/2; ss=sqrt(1-x*x); p=ss; 
		if(n==1) ss=0.0;
		for(k=1; k<=l; k++)
		{
			p=p*2*k*(1-x*x)/(2*k+1.0); ss=ss+p;
		}
		f=(asin(x)+x*ss)*2/PI;
	}
	return 1-f;
}


/*	To generate uniformly distributed random numbers in interval (0,1)

	SEED : (input/output) is a real value used as the seed
		It should be positive during initial call and
		should not be modified between different calls

	Required functions : None
*/

#include <math.h>

double ran1(double *seed)

{
	double am=2147483648e0, a=45875e0, ac=453816693e0, an=2147483647e0;

	*seed=fmod((*seed)*a+ac,am);
	return (*seed)/an;
}
	
